### ISSUE TITLE
It isn't possible to drag-and-drop selected text

### ISSUE THREAD

**Issue by izuzak** [atom/atom#1909]
I was sure an issue was already opened for this, but can't find it now.

Halp ticket:
- support/5124ebc4cfa211e3871387e11a229b02

Would be cool if it was possible to select some text and then drag-and-drop it to some other place.

**Comment by scottbedard** [comment]
+1. I'm with JayQuerido, this is the only feature preventing me from leaving Sublime.

**Comment by CarbonFactory** [comment]
+1 
Badly missing this feature

**Comment by apt-develop** [comment]
+1

I need this feature.

**Comment by Thijs5** [comment]
+1
Every other text-editor has this behavior.

**Comment by clemorphy** [comment]
+1
I absolutely need this

**Comment by mark-hahn** [comment]
**It is now possible to drag and drop text.**

https://atom.io/packages/drag-drop-text

Besides the normal drag/drop it supports cut/copy/paste using the mouse and no keys, not even modifier keys.

This is brand new so please get me feedback on the design and of course report bugs.

**Comment by markchagers** [comment]
First-off a big thank-you for implementing this feature!
I do have a few remarks though:
- In the current implementation it is somewhat confusing: sometimes the selected text is copied, sometimes it is moved or even just removed. I can't really tell what is going to happen.
- when I make a selection and then click and hold on the selected text, it hides the selection, although it seems to be retained invisibly, when you hold and drag the previously selected text gets a red outline.
- I find it easier to use a modifier key to distinguish between regular dragging (moving) and copy-dragging the way it's implemented in most text editors that have this feature.

**Comment by mark-hahn** [comment]
Thanks very much for the comments.  I have worried about the same things.  Let me explain how it ended up this way and then maybe people can help me redesign the UX.

>  it is somewhat confusing: sometimes the selected text is copied, sometimes it is moved or even just removed.

This is due to my getting carried away with features.  While working on it I discovered it would be easy to replace `ctrl-x`, `ctrl-c`, and `ctrl-v` with mouse-only actions.  I am a heavy mouse-user and I have already gotten hooked on this feature.  Having said that, I will investigate adding a setting to remove those features, assuming it makes things simpler.

> sometimes the selected text is copied

There is a copy to the clipboard when and only when the red box appears.

> sometimes it is moved

If you continue to hold the mouse down after the red box appears then the text will be deleted.  That is the only difference between a copy and move.  The timing for the delays are set in settings.  Maybe starting off with a long period will make the behavior more understandable.

>  or even just removed.

If you don't move the mouse after you do the long hold and cut then there is no drop.  This gives you the straight `ctrl-x`.  I'm not sure what I can do about this.

> when I make a selection and then click and hold on the selected text, it hides the selection, although it seems to be retained invisibly, when you hold and drag the previously selected text gets a red outline.

Yes.  When you click on a pre-existing selection the text disappears but that doesn't affect the functionality, as you can see by the red box appearing at the right time and place anyway.  This annoys me also.

This weird behavior is due to the decision I made early on to not "interfere" with any existing Atom behavior.  In Atom, clicking on a selection kills it immediately on the mousedown.  The Atom behavior is non-standard.  Just now while typing in this textarea I tried and the selection is not cleared until after the mouse has been lifted.  I will attempt to fix Atom but that may be really hard.

> I find it easier to use a modifier key to distinguish between regular dragging (moving) and copy-dragging the way it's implemented in most text editors that have this feature.

Yes that should be an option also.  I like not having to move to the keyboard but I'm weird.  There is a problem though.  `Alt` is often a modifier for rectangular selection and `Ctrl` is definitely used for multiple selections.  I will add a setting for using `Alt`.

So in summary I will investigate each of ...

- Making  `ctrl-x`, `ctrl-c`, and `ctrl-v` a setting
- Fixing Atom to not clear selection until `mouseup`
- Making the `Alt` key an option to specify copying instead of using delays

Please keep the comments coming ...

**Comment by pbihler** [comment]
First: Thanks for your initiative on the topic!

From the UX expert perspective, you should strive for consistency as default (First of [Shneiderman's "Eight Golden Rules of Interface Design"](http://faculty.washington.edu/jtenenbg/courses/360/f04/sessions/schneidermanGoldenRules.html)).

If I remember correctly, the default behavior for drag'n' drop in editors, you have:
* Drag'n'Drop for block move
* Drag'n'<kbd>modifier</kbd>-Pressed'n'Hold'n'Drop for copy, with modifier = <kbd>cmd</kbd> on Mac an <kbd>ctrl</kbd> on Windows (Don't know about Linux, probably also <kbd>ctrl</kbd>). 
* No "Cut" possible (you would need a trash-bin icon or so to make the mapping clear (1&2 of [Nielsens UI Heuristics](http://www.nngroup.com/articles/ten-usability-heuristics/))
* You should not use the Clipboard, Drag'n'Drop is independent of that.

You can add an settings option "Enable Mouse-Only-Mode" for your current behavior and explain it (like above) in the README, since it's hard to make the system status clear for that mode.

**Comment by varemenos** [comment]
+1 to @pbihler 's thoughts

**Comment by mark-hahn** [comment]
>  the default behavior for drag'n' drop in editors,

Yes that sounds correct for when my crazy cut/copy/paste features are disabled.  The behavior is going to be very different with that setting set.  It should almost be different packages.

**Comment by mark-hahn** [comment]
I'm going to make it a separate package.  I hope it doesn't confuse people.  The new one will be `simple-drag-drop-text` but I'll keep the old name the same and make the difference clear in the one-line description.

**Comment by mark-hahn** [comment]
I think I've got it now.  Let me know if it still has any weirdness.  It is at https://atom.io/packages/simple-drag-drop-text.

A discussion on the forum is at https://discuss.atom.io/t/announce-simple-drag-drop-text/15181.

**Comment by mark-hahn** [comment]
OK, **now** I've finally got it.   I removed the delay and red line in the `simple-drag-drop` package and now it behaves exactly like other editors.  Thanks to @btleffler for bitching until I got it right.  https://discuss.atom.io/t/announce-simple-drag-drop-text/15181

See the new version at https://atom.io/packages/simple-drag-drop-text.

**Comment by adampope** [comment]
Seems to be perfect!  Thanks @mark-hahn :)

**Comment by adunning** [comment]
Thanks for that package! That said, I would still like to see Atom adopt the standard OS X service for dragging text: it should use fewer resources (`simple-drag-drop` adds 45ms to the start time), and it's also much better from a usability perspective, since one can see the text below the cursor as it is being dragged.

**Comment by robsonsobral** [comment]
To not be able to drag text is something pretty weird for me. I can't even drag text from the browser and drop it on the editor.

The ironic part is that even IE allows it.

**Comment by Zireael07** [comment]
@robsonsobral: Have you seen the simple-drag-drop-text package?

**Comment by robsonsobral** [comment]
@Zireael07 , I did, yes. However, I don't think an addon is enough. Of course, is better than nothing and I thankful to @mark-hahn to code it, but I think it should be part of the core or at least an official package. A calculator is a feature for a package, or git integration, which Atom already has; not something that even WordPad does. This functionality is so common that at the first moment I though it was lacking due to a bug. From a UX perspective, it is a bug.

Plus, the package doesn't allow us to drag and drop from or to outside Atom.

**Comment by SoullessWaffle** [comment]
@robsonsobral I opened [an issue](https://github.com/mark-hahn/simple-drag-drop-text/issues/9) on @mark-hahn's package about this.
I 100% agree that this should be part of the core.

**Comment by robsonsobral** [comment]
@SoullessWaffle , I'm already subscribed to the issue you opened.

Thank you.

**Comment by devios1337** [comment]
I would definitely consider missing a standard feature like drag and drop a bug. @mark-hahn's plugin is definitely a big help, so thanks to you Mark for giving us something to use in the meantime. But the fact is this is a hack for something that really should be supported by core. The biggest shortcoming right now is the fact that Atom is completely ignorant of drags originating from or destined to another application.

**Comment by robsonsobral** [comment]
One more:

- we can't drag from one file to another not even using @mark-hahn plugin;

**Comment by mark-hahn** [comment]
> Atom is completely ignorant of drags originating from or destined to
another application.

As far as I know this is impossible in linux.

**Comment by dennify** [comment]
+1 to implement this to core functionality

**Comment by axefrog** [comment]
@mark-hahn your package is excellent and I find it very useful, so thanks for taking the time to implement it!

This should definitely be built into Atom core though. It's kind of a basic, standard editing experience that most other text editors have, and which is surprising to find doesn't exist in Atom.

**Comment by DeadLion** [comment]
+1

**Comment by lotusms** [comment]
This exists here [https://atom.io/packages/simple-drag-drop-text]( https://atom.io/packages/simple-drag-drop-text)

**Comment by axefrog** [comment]
@lotusms Please actually read the preceding messages before posting.

**Comment by lotusms** [comment]
@axefrog nah. Too much to read. But thanks!

**Comment by lee-dohm** [comment]
@lotusms @axefrog We have a [Code of Conduct](https://github.com/atom/atom/blob/master/CODE_OF_CONDUCT.md) that includes, among other things, an expectation to keep things positive and respectful. Keep this in mind in your future participation. You may both consider this an official warning.

**Comment by paulbhartzog** [comment]
Just wondering if this is being addressed.  I don't care about dragging text inside of the app, but not being able to drag text to the Desktop or to other apps is crucial.  Thx.

**Comment by GeorgeWL** [comment]
Could we add this into the core package list then?

I just feel if a package is used by a huge enough portion it doesn't make sense for it to have to be manually installed.

https://atom.io/packages/simple-drag-drop-text

I'd even suggest having a setting in the editor options to dropdown between the two often used packages of that and 

https://atom.io/packages/drag-drop-text

**Comment by bronson** [comment]
You want to merge two competing packages into core, and have a dropdown to select which one is used?

**Comment by GeorgeWL** [comment]
ah true enough you've got a point there...

Just one would make more sense.

**Comment by 50Wliu** [comment]
We currently have _too many_ core packages to maintain, so adding another one to the list is a bit out of the question at this point in time.

Also, Linter was never a core package, at least since Atom's open-source release.

**Comment by GeorgeWL** [comment]
I still think this is a case of,  if literally every other text editor, even basic notepad has it,  that probably means that it's worth adding to core no matter the cost of maintenance. 

Simple Drag Drop though,  if you look at the package, is incredibly minimal.

**Comment by davidak** [comment]
I still can reproduce this issue.

Atom    : 1.24.0
Electron: 1.6.16
Chrome  : 56.0.2924.87
Node    : 7.4.0

apm  1.18.12
npm  3.10.10
node 6.9.5 x64
atom 1.24.0
python 
git 2.15.0

 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.86, NixOS, 17.09.3202.9a8344a7a72 (Hummingbird)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.0`
 - channels(davidak): `""`
 - channels(root): `"nixos-17.09, unstable, nixos-hardware"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

Steps to reproduce: select text (from this issue) and drag onto text file in atom... nothing happens

**Comment by bronson** [comment]
Oh good. I never dnd text, but I've seen so many other people do it that this really seems worth fixing.

**Comment by davidak** [comment]
This is still not implemented.

Atom 1.30.0

**Comment by lee-dohm** [comment]
Thanks everyone for your feedback.

After reviewing this with the team, the maintainers have decided that we aren't going to be taking this on. In addition, due to the complexity of the requirements of the feature, we won't be entertaining a pull request to add this either.
